cpp/include/ucxx/memory_handle.h

```diff
@@ -0,0 +1,162 @@
+  MemoryHandle(const MemoryHandle&)            = delete;
+  MemoryHandle& operator=(MemoryHandle const&) = delete;
+  MemoryHandle(MemoryHandle&& o)               = delete;
+  MemoryHandle& operator=(MemoryHandle&& o)    = delete;
+
+  /**
+   * @brief Constructor for `shared_ptr<ucxx::MemoryHandle>`.
+   *
+   * The constructor for a `shared_ptr<ucxx::MemoryHandle>` object, mapping a memory buffer
+   * with UCP to provide RMA (Remote Memory Access) to.
+   *
+   * The allocation requires a `size` and a `buffer`.  The `buffer` provided may be either
+   * a `nullptr`, in which case UCP will allocate a new memory region for it, or an already
+   * existing allocation, in which case UCP will only map it for RMA and it's the caller's
+   * responsibility to keep `buffer` alive until this object is destroyed. When the UCP
+   * allocates `buffer` (i.e., when the value passed is `nullptr`), the actual size of the
+   * allocation may be larger than requested, and can later be found calling the `getSize()`
+   * method, if a preallocated buffer is passed `getSize()` will return the same value
+   * specified for `size`.
+   *
+   * @code{.cpp}
+   * // `context` is `std::shared_ptr<ucxx::Context>`
+   * // Allocate a 128-byte buffer with UCP.
+   * auto memoryHandle = context->createMemoryHandle(128, nullptr);
+   *
+   * // Equivalent to line above
+   * // auto memoryHandle = ucxx::createMemoryHandle(context, 128, nullptr);
+   *
+   * // Map an existing 128-byte buffer with UCP.
+   * size_t allocationSize = 128;
+   * auto buffer = new uint8_t[allocationSize];
+   * auto memoryHandleFromBuffer = context->createMemoryHandle(
+   *    allocationSize * sizeof(*buffer), reinterpret_cast<void*>(buffer)
+   * );
+   *
+   * // Equivalent to line above
+   * // auto memoryHandleFromBuffer = ucxx::createMemoryHandle(
+   * //    context, allocationSize * sizeof(*buffer), reinterpret_cast<void*>(buffer)
+   * // );
+   * @endcode
+   *
+   * @throws ucxx::Error if either `ucp_mem_map` or `ucp_mem_query` fail.
+   *
+   * @param[in] context parent context where to map memory.
+   * @param[in] size    the minimum size of the memory allocation
+   * @param[in] buffer  the pointer to an existing allocation or `nullptr` to allocate a
+   *                    new memory region.
+   *
+   * @returns The `shared_ptr<ucxx::MemoryHandle>` object
+   */
+  friend std::shared_ptr<MemoryHandle> createMemoryHandle(std::shared_ptr<Context> context,
+                                                          const size_t size,
+                                                          void* buffer);
+
+  ~MemoryHandle();
+
+  /**
+   * @brief Get the underlying `ucp_mem_h` handle.
+   *
+   * Lifetime of the `ucp_mem_h` handle is managed by the `ucxx::MemoryHandle` object and
+   * its ownership is non-transferrable. Once the `ucxx::MemoryHandle` is destroyed the
+   * memory is unampped and the handle is not valid anymore, it is the user's responsibility
```

Comment:
```suggestion
   * memory is unmapped and the handle is not valid anymore, it is the user's responsibility
```

Reply:
Thanks for catching that, a bit disappointed codespell didn't catch it though.